### ISSUE TITLE
User is not informed if there is a clickable element in the cells of detailslist component while using narrator/nvda.

### ISSUE THREAD

**Issue by SriNandiniG** [OfficeDev/office-ui-fabric-react#8493]
## Codepen demonstrating the issue
 https://codepen.io/srinandini/pen/EMzMZE ## Actual behavior:

## Expected behavior: 
The narrator/nvda has to inform the user if there is clickable element in the detailslist

## Documentation describing expected behavior

In the codepen demonstation, there is a dialog box that shows up upon clicking elements in the detailslist. The narrator/nvda does not mention button/link which tells the users that there is on click functionality for the button.

**Comment by Vitalius1** [comment]
@KevinTCoughlin @natalieethell you both dealt with some Narrator issues in DetailsList and I was wondering if any practical knowledge from that can be used in debugging this issue 😄

**Comment by KevinTCoughlin** [comment]
@SriNandiniG what should the rows announce? Note that we focus on Narrator + Edge compatibility.

Does the same issue occur with that combination?

Also might be worth noting that in your example the dialog box only occurs on 'click'. Keyboard invocation invokes `window.alert`.

**Comment by SriNandiniG** [comment]
Hi Kevin,

I have checked in issue in Narrator + Edge combination. The narrator announces the text present in the cells. I wanted to check with you if it’s possible for the user to know that the elements in cells are clickable.
Since the role of the elements in the cells, is set automatically for all the cells in details list, the narrator is not informing the user regarding the on click function available for the element. It can be helpful if we can modify the role of the cells so the narrator informs the same.
Thank you for your quick response.

Regards,
Nandini

**Comment by Vitalius1** [comment]
Here is an updated [codepen](https://codepen.io/vitalius1/pen/OqerJN) which would help to debug what Narrator is announcing. (need to setup my Windows environment first 😄 )
I can see 3 scenarios:
1. When tabbing to the entire row because we have an `onItemInvoked` prop on it it looks like the row it's actionable so I would expect the Narattor to let the user know about that.
2. When tabbing to the entire row and the Narattor starts reading all the cells <b>here is where I wonder what the Narattor should read out. Is there even a simple way to make Narattor announce for each cell if it's read-only or actionable. What about the Scan mode?</b>
3. When arrowing to the name (which is essentially a button now) I would expect the Narrator saying something about that.

**Comment by natalieethell** [comment]
@SriNandiniG so to clarify, you'd like for Narrator to announce when something has an onclick event? As Kevin touched on above, folks who use Narrator typically use keyboard navigation and not a mouse so I'm just a little confused as to the use case here. 

I might be missing something - is there a scenario in which someone using Narrator would use a mouse? Or is the solution here informing the user when there is a keyboard action they can take? 

Lastly, what is the exact read-out you're hoping to get here from assistive technology?

**Comment by SriNandiniG** [comment]
Hi,

Since the dialog box opens when we navigate to the row with keyboard (and press enter key), If the narrator mentions the role like link/cickable element etc, so  the user would know that there is a dialog box which opens on pressing enter when we reach the row.

Thanks,
Nandini

**Comment by natalieethell** [comment]
@SriNandiniG thanks for the clarification! 

So it looks like now, when you navigate within a row to the gridcell that has a button, Narrator will announce "button". The dialog only opens when focused on that button, not on the row itself—right?

When you navigate to the row itself, it reads out each cell value, but not the role/type of each cell. Here's an example from Vitalie's [CodePen](https://codepen.io/vitalius1/pen/OqerJN):

**Lorem ipsum.pub, Dolor Sit, 98KB, non-selected**

So when navigating row by row, you would like Narrator to read out each cell value along with each role/type? Something like the following?

**Lorem ipsum.pub button, Dolor Sit, 98KB, non-selected**

**Comment by SriNandiniG** [comment]
Hi,

Based on the CodePen<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fcodepen.io%2Fvitalius1%2Fpen%2FOqerJN&data=02%7C01%7Cv-srigan%40microsoft.com%7Cda9150fda03e4446c0e408d6b85556ef%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636899074003785119&sdata=M1SIrTfejCX4GZr5kb595vMQzQzUa21c1dIFXVExiaA%3D&reserved=0> you have given, now the dialog opens only when focused on the button, we do not have any issue with dialog box opening upon clicking anywhere on the row, as it used to be earlier.

While navigating row by row, now the narrator reads out each cell value along with each role/type,
Lorem ipsum.pub button, Dolor Sit, 98KB, non-selected,

Yes, I wanted to clarify if this would be possible that narrator reads out each cell value along with role/type so  the user would know that there is a dialog box which opens on pressing enter when we reach the row.

Thanks,
Nandini

**Comment by natalieethell** [comment]
@SriNandiniG got it, thanks for the clarification! 

So, to obtain the behavior you're looking for, it seems like a fix needs to be implemented in the screen reader, not in the DetailsList control.

When focused on a row, the screen reader just reads out the content (text) in each gridcell in that row. In order to find out more information about each cell, you'll need to navigate to each individual cell in that row.

As for letting the user know when a row is actionable (such as, hitting enter opens a dialog by passing a callback to the `onItemInvoked` prop), this seems like something custom you could add to the aria label of the row itself. @Vitalius1 and I think this would be the best way to support this scenario. We talked about conditionally adding a new role, like a 'button' to the row element, but there's no way really for the DetailsList to know what was passed into the `onItemInvoked` prop. It could be a console.log, it could open a dialog, etc. 

In summary, not reading out the role of each element in a row seems more like a screen reader issue than a Fabric issue.

Let us know if that answers your question!

**Comment by SriNandiniG** [comment]
Thanks for your efforts! I have raised the same issue to NVDA, to check if the ask is possible

**Comment by Vitalius1** [comment]
@SriNandiniG will be closing this issue and feel free to comment back if you get any resolution from screen readers' side.

**Comment by KeiraYang** [comment]
How can I not receive the email? I've unsubscribe it but continue to get it. 


谢谢～
---Original---
